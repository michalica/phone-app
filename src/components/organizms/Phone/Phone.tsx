import React, { ChangeEvent, useState } from 'react';
import styled from '@emotion/styled';
import DefaultInput from '../../molecules/DefaultInput';
import Suggestions from '../../molecules/Suggestions';
import Words from '../../molecules/Words';
import Button from '../../atoms/Button';

export const defaultProps = {
  Wrapper: styled.div`
    border: 2px solid #00a991;
    max-width: 300px;
    height: 400px;
    border-radius: 20px;
    background: ${(props) => props.color};
  `,
  Display: styled.div`
    border: 1px solid #cdcdcd;
    margin: 15px;
    display: flex;
    flex-direction: column;
    border-radius: 10px;
    align-items: center;
    padding: 10px;
  `,
  SuggestionsWrapper: styled.div`
    height: 120px;
    margin-top: 10px;
  `,
  WordsWrapper: styled.div`
    height: 100px;
    display: flex;
  `,
  Row: styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
  `,
  Keyboard: styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
  `,
  Speaker: styled.div`
    border: 2px solid black;
    width: 20px;
    margin: 0px auto;
    margin-top: 10px;
  `,
  Input: DefaultInput,
  SuggestionComponent: Suggestions,
  WordsComponent: Words,
};

type Props = Readonly<
  typeof defaultProps & {
  suggestions: string[];
  words: string[];
  onType?(newValue: string): void;
  value?: string;
  setValue?: (value: string) => void;
  color?: string;
  setColor?: (value: string) => void;
  add?: (addValue: string, props: Partial<Props>) => void;
  getRandomColor?: () => string;
}
>;
type Children = (props: Props) => JSX.Element;

const Phone = ({ children, ...props }: { children: Children } & Props) => {
  const [value, setValue] = useState('');
  const [color, setColor] = useState('white');

  return children({ color, setColor, value, setValue, ...props });
};

Phone.defaultProps = {
  ...defaultProps,
  getRandomColor: (): string => {
    const letters = '0123456789ABCDEF';
    let color = '#';
    for (let i = 0; i < 6; i++) {
      color += letters[Math.floor(Math.random() * 16)];
    }
    return color;
  },
  add: (
    addValue: string,
    { value, setValue, onType, setColor, getRandomColor }: Partial<Props>
  ): void => {
    setValue!(`${value + addValue}`);
    onType?.(`${value + addValue}`);
    setColor!(getRandomColor!());
  },
  children: ({
    Wrapper,
    Input,
    onType,
    SuggestionComponent,
    suggestions,
    SuggestionsWrapper,
    Display,
    Keyboard,
    Row,
    setValue,
    value,
    Speaker,
    add,
    getRandomColor,
    setColor,
    color,
  }: Props) => (
    <Wrapper color={color}>
      <Speaker />
      <Display>
        <>
          <div>
            Try <b>236</b>:
          </div>
          <SuggestionsWrapper>
            <SuggestionComponent
              suggestions={suggestions}
              numberOfSuggestions={3}
            />
          </SuggestionsWrapper>
          <Input
            onChange={(e: ChangeEvent<HTMLInputElement>) => {
              const eValue = e.target.value;
              setValue?.(eValue);
              onType?.(eValue);
              setColor?.(getRandomColor!());
            }}
            value={value}
          />
        </>
      </Display>

      <Keyboard>
        <Row>
          <Button disabled>1</Button>
          <Button
            onClick={() =>
              add!('2', { value, onType, setValue, setColor, getRandomColor })}
          >
            2
          </Button>
          <Button
            onClick={() =>
              add!('3', { value, onType, setValue, setColor, getRandomColor })}
          >
            3
          </Button>
        </Row>
        <Row>
          <Button
            onClick={() =>
              add!('4', { value, onType, setValue, setColor, getRandomColor })}
          >
            4
          </Button>
          <Button
            onClick={() =>
              add!('5', { value, onType, setValue, setColor, getRandomColor })}
          >
            5
          </Button>
          <Button
            onClick={() =>
              add!('6', { value, onType, setValue, setColor, getRandomColor })}
          >
            6
          </Button>
        </Row>
        <Row>
          <Button
            onClick={() =>
              add!('7', { value, onType, setValue, setColor, getRandomColor })}
          >
            7
          </Button>
          <Button
            onClick={() =>
              add!('8', { value, onType, setValue, setColor, getRandomColor })}
          >
            8
          </Button>
          <Button
            onClick={() =>
              add!('9', { value, onType, setValue, setColor, getRandomColor })}
          >
            9
          </Button>
        </Row>
      </Keyboard>
    </Wrapper>
  ),
};

export default Phone;
