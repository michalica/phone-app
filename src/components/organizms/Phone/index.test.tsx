import React from 'react';
import { render } from '@testing-library/react';
import Phone from './Phone';

describe('snapshot tests', () => {
  it('testing rendered tree', () => {
    const { container } = render(<Phone />);
    expect(container).toMatchSnapshot();
  });
});
