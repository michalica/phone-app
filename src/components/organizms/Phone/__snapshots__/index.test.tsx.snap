// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing testing rendered tree 1`] = `
.emotion-18 {
  border: 2px solid #00a991;
  max-width: 300px;
  height: 400px;
  border-radius: 20px;
  background: white;
}

.emotion-0 {
  border: 2px solid black;
  width: 20px;
  margin: 0px auto;
  margin-top: 10px;
}

.emotion-4 {
  border: 1px solid #cdcdcd;
  margin: 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-radius: 10px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
}

.emotion-2 {
  height: 120px;
  margin-top: 10px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-3 {
  border-top-style: hidden;
  border-right-style: hidden;
  border-left-style: hidden;
  border-color: #cdcdcd;
  outline: none;
  height: 30px;
  font-family: 'Roboto',sans-serif;
  text-align: center;
}

.emotion-17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.emotion-5 {
  background: none;
  border: none;
  color: white;
  background: #00a991;
  padding: 15px;
  cursor: pointer;
  border: 1px solid white;
}

<div>
  <div
    class="emotion-18"
    color="white"
  >
    <div
      class="emotion-0"
    />
    <div
      class="emotion-4"
    >
      <div>
        Try 
        <b>
          236
        </b>
        :
      </div>
      <div
        class="emotion-2"
      >
        <div
          class="emotion-1"
        />
      </div>
      <input
        aria-label="numbers"
        class="emotion-3"
        value=""
      />
    </div>
    <div
      class="emotion-17"
    >
      <div
        class="emotion-8"
      >
        <button
          class="emotion-5"
          disabled=""
        >
          1
        </button>
        <button
          class="emotion-5"
        >
          2
        </button>
        <button
          class="emotion-5"
        >
          3
        </button>
      </div>
      <div
        class="emotion-8"
      >
        <button
          class="emotion-5"
        >
          4
        </button>
        <button
          class="emotion-5"
        >
          5
        </button>
        <button
          class="emotion-5"
        >
          6
        </button>
      </div>
      <div
        class="emotion-8"
      >
        <button
          class="emotion-5"
        >
          7
        </button>
        <button
          class="emotion-5"
        >
          8
        </button>
        <button
          class="emotion-5"
        >
          9
        </button>
      </div>
    </div>
  </div>
</div>
`;
